src/samarium/shell.py

```diff
@@ -9,28 +16,129 @@
+    def __init__(self, debug: bool = False) -> None:
+        self.registry = Registry(globals())
+        self.prompt = IN
+
+        self.messages = [dahlia(f"&3Samarium {__version__}"), self.prompt]
+
+        self.statement = ""
+        self.length = 1
+
+        self.io = FakeIO()
+        self.stdout = sys.stdout.write
+        self.stderr = sys.stderr.write
+
+        self.debug = debug
+
+    def execute(self, code: str) -> None:
+        sys.stdout.write = sys.stderr.write = self.io.write
+        run(
+            code,
+            self.registry,
+            debug=self.debug,
+            load_template=False,
+            quit_on_error=False,
+        )
+        sys.stdout.write = self.stdout
+        sys.stderr.write = self.stderr
+
+    def render(self, window: FullscreenWindow) -> None:
+        window.render_to_terminal(
+            self.messages, (len(self.messages) - 1, len(clean_ansi(self.messages[-1])))
+        )
+
+    def run(self):
+        with FullscreenWindow(hide_cursor=False) as window:
+            with Input() as input_gen:
+                self.render(window)
+                for e in input_gen:
+                    if self.handle(e) is True:
+                        break
+                    self.render(window)
+
+    def handle(self, e: Event | str | None) -> bool | None:
+        if e == "<Ctrl-d>":
+            return True
+        elif e == "<Ctrl-c>":
+            self.length = 1
+            self.prompt = IN
+            self.messages.append(self.prompt)
+        elif e == "<SPACE>":
+            self.messages[-1] += " "
+        elif e == "<TAB>":
+            self.messages[-1] += "    "
+        elif e == "<Ctrl-l>":
+            self.messages = [self.messages[-1]]
+        elif e == "<BACKSPACE>":
+            curr = self.messages[-1]
+            if len(curr) + 4:
+                self.messages[-1] = self.prompt + curr[len(self.prompt) : len(curr) - 1]
+        elif e == "<Ctrl-j>":
+            if self.on_return() is False:
+                return None
+        elif isinstance(e, PasteEvent):
+            self.on_paste(e)
+        elif (
+            isinstance(e, Event) or e is None or (e.startswith("<") and e.endswith(">"))
+        ):
+            return None
+        else:
+            self.messages[-1] += e
+
+    def format_statement(self, statements: list[str]) -> str:
+        statements[0] = statements[0][len(IN) :]
+
+        statements = [
+            i[len(INDENT) :] if i.startswith(INDENT) else i for i in statements
+        ]
+
+        return "\n".join(statements)
+
+    def on_return(self):
+        statement = self.format_statement(self.messages[-self.length :])
+        if not statement:
+            return False
+        statement += " ;"
+        error, _ = match_brackets(tokenize(statement))
+
+        if not error:
+            self.execute(statement)
+
+            if self.io.output:
+                self.messages.extend("".join(self.io.output).split("\n"))
```

Comment:
```suggestion
                self.messages.extend("".join(self.io.output).splitlines())
```

Reply:
this is intentional. it is not splitting lines if a file but splitting the newlines that written to stdout/stderr.

Reply:
and... ?

Reply:
iirc splitlines can split on things other than `\n`